### PR TITLE
Scaffold smart contract event storage

### PR DESCRIPTION
**Pull request by neekolas**
## tl;dr

- Adds function to run log streamer and store events
- Creates flexible interface for deserializing and storing events
- Updates mockery config and moves all mocks to a dedicated folder

## AI Assisted summary
- **blockchain/interfaces**: Remove `Stop()` method from `LogStreamer` interface, add custom logger name to `RpcLogStreamer`.
- **blockchain/rpcLogStreamer.go**: Add buffered channel to `ListenForContractEvent` to improve event handling.
- **indexer/indexer.go**: Add new file for starting the indexer service with configuration for message contract events.
- **storer**: Implement error handling for the log storer, including support for retryable errors. Add mock for `LogStorer`.
- **Refactoring**: Move existing mock generation to the 'mocks' package for better code organization.
- **server/options.go**: Integrate `ContractsOptions` into the server options for configuration.

### CODE REVIEW

**Comment by neekolas** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/xmtp/xmtpd/105?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#105** <a href="https://app.graphite.dev/github/pr/xmtp/xmtpd/105?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#103** <a href="https://app.graphite.dev/github/pr/xmtp/xmtpd/103?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#102** <a href="https://app.graphite.dev/github/pr/xmtp/xmtpd/102?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @neekolas and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: pkg/indexer/indexer.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+			err = logStorer.StoreLog(ctx, event)
+			if err != nil {
+				logger.Error("error storing log", zap.Error(err))
+				if err.ShouldRetry() {
+					time.Sleep(100 * time.Millisecond)
+					continue Retry
+				}
+			}
+			logger.Info("Stored log", zap.Uint64("blockNumber", event.BlockNumber))
+			break Retry
```
Comment (richardhuaaa) [inline pkg/indexer/indexer.go:59]: Are the continue/break labels needed here? It seems like we're only breaking out of one loop

File: pkg/indexer/indexer.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+		for {
+			err = logStorer.StoreLog(ctx, event)
+			if err != nil {
+				logger.Error("error storing log", zap.Error(err))
+				if err.ShouldRetry() {
+					time.Sleep(100 * time.Millisecond)
+					continue Retry
+				}
+			}
+			logger.Info("Stored log", zap.Uint64("blockNumber", event.BlockNumber))
```
Comment (richardhuaaa) [inline pkg/indexer/indexer.go:58]:
```suggestion
			} else {
				logger.Info("Stored log", zap.Uint64("blockNumber", event.BlockNumber))
			}
```

File: pkg/indexer/storer/error.go [hunk 3]
```diff
@@ -0,0 +1,26 @@
+package storer
+
+type LogStorageError interface {
+	error
+	ShouldRetry() bool
```
Comment (richardhuaaa) [inline pkg/indexer/storer/error.go:5]: Love that retries are built in from the beginning

**Review by richardhuaaa — APPROVED** [review]

File: pkg/indexer/indexer.go (hunk 1 above)
Comment (neekolas) [reply pkg/indexer/indexer.go:59]: Not needed. I kinda like them for clarity, but that's definitely more a question of style than anything else.

File: pkg/indexer/indexer.go (hunk 2 above)
Comment (neekolas) [reply pkg/indexer/indexer.go:58]: Good catch
